File: permutation.py
from operation import Operation


def permutation(size, n, op, result):
    if op == Operation.Divide:
        maxes = list(filter(lambda x: x % result == 0 and x > 0, range(size+1)))
        mins = [m//result for m in maxes]
        p = []
        for x in range(len(maxes)):
            p.append([maxes[x], mins[x]])
            p.append([mins[x], maxes[x]])
        return p

    if op == Operation.Subtract:
        maxes = list(filter(lambda x: x > result, range(size+1)))
        mins = [m-result for m in maxes]
        p = []
        for x in range(len(maxes)):
            p.append([maxes[x], mins[x]])
            p.append([mins[x], maxes[x]])
        return p

    if op == Operation.Add:
        return pe_add(size, result, n)

    if op == Operation.Multiply:
        return pe_multiply(size, result, n)


def pe_add(ma, res, n):
    if n == 1:
        return [[res]]
    t = []
    for x in range(1, ma+1):
        if (res-x)/(n-1) <= ma and res-x >= n-1:
            for ss in pe_add(ma, res-x, n-1):
                t.append(ss + [x])
    return t


def pe_multiply(ma, res, n):
    if n == 1:
        return [[res]]
    t = []
    factors = list(filter(lambda x: res % x == 0, range(1, ma+1)))
    for x in factors:
        if res//x <= ma**(n-1):
            for ss in pe_multiply(ma, res//x, n-1):
                t.append(ss + [int(x)])
    return t

File: KenKen Solver/permutation.swift
//
//  permutation.swift
//  KenKen Solver
//
//  Created by Emil Bach Mikkelsen on 4/14/22.
//


func permutations(boardSize: Int, groupSize: Int, operation: Operation, result: Int) -> [[Int]] {
    if (operation == Operation.divide) {
        var maxes: [Int] = [];
        var mins: [Int] = [];
        
        for x in 1..<boardSize+1 {
            if (x % result == 0) {
                maxes.append(x);
                mins.append(x / result);
            }
        }
        
        var p: [[Int]] = [];
        
        for (ma, mi) in zip(maxes, mins) {
            p.append([ma, mi]);
            p.append([mi, ma]);
        }
        
        return p;
    }
    
    if (operation == Operation.subtract) {
        var maxes: [Int] = [];
        var mins: [Int] = [];
        
        for x in 1..<boardSize+1 {
            if (x > result) {
                maxes.append(x);
                mins.append(x - result);
            }
        }
        var p: [[Int]] = [];
        for (ma, mi) in zip(maxes, mins) {
            p.append([ma, mi]);
            p.append([mi, ma]);
        }
        return p;
    }
    
    if (operation == Operation.add) {
        return peAdd(boardSize: boardSize, result: result, groupSize: groupSize);
    }
        
    return peMultiply(boardSize: boardSize, result: result, groupSize: groupSize);
}


func peAdd(boardSize: Int, result: Int, groupSize: Int) -> [[Int]] {
    if (groupSize == 1) {
        return [[result]];
    }
    var t: [[Int]] = [];
    for x in 1..<boardSize+1 {
        if ((result - x) / (groupSize - 1) <= boardSize && result - x >= groupSize - 1) {
            for ss in peAdd(boardSize: boardSize, result: result - x, groupSize: groupSize - 1) {
                t.append(ss + [x])
            }
        }
    }
    
    return t;
}
    

func peMultiply(boardSize: Int, result: Int, groupSize: Int) -> [[Int]] {
    if (groupSize == 1) {
        return [[result]];
    }
    var t: [[Int]] = [];
    var factors: [Int] = [];
    for x in 1..<boardSize+1 {
        if (result % x == 0) {
            factors.append(x);
        }
    }
    for x in factors {
        if (result / x) <= boardSize^(groupSize - 1) {
            for ss in peMultiply(boardSize: boardSize, result: result / x, groupSize: groupSize - 1) {
                t.append(ss + [x])
            }
        }
    }
    
    return t;
}
  
